File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/fs/DirectoryPresenceRule.java
package com.ririnto.sinon.harness.rules.fs;

import com.ririnto.sinon.harness.rules.HarnessCheckHelper;
import com.ririnto.sinon.harness.rules.HarnessCheckRule;
import com.ririnto.sinon.harness.core.RuleContext;
import com.ririnto.sinon.harness.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.util.Collection;

/**
 * Rule that requires specified directories to exist in the project.
 */
public enum DirectoryPresenceRule implements HarnessCheckRule {
    INSTANCE;

    private static final String CATEGORY = "directoryPresence";

    @Override
    public String category() {
        return "directoryPresence";
    }

    @Override
    public boolean applies(RuleContext ctx) {
        return ctx.manifest().isEnabled(CATEGORY);
    }

    @Override
    public Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException {
        final JsonNode manifest = ctx.manifest().raw();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return HarnessCheckHelper.extractPaths(manifest.get(CATEGORY).get("parameters").get("paths")).stream()
                .filter(path -> !HarnessCheckHelper.isSafeDirectory(ctx.root(), ctx.root().resolve(path)))
                .map(path -> Finding.of(severity, CATEGORY, "missing directory: " + path))
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/fs/DirectoryPresenceRule.kt
package com.ririnto.sinon.harness.rules.fs

import com.ririnto.sinon.harness.ast.AstFinding
import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import com.ririnto.sinon.harness.core.HarnessCheck
import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.core.Severity
import com.ririnto.sinon.harness.rules.HarnessCheckRule
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isDirectory
import kotlin.io.path.isSymbolicLink

/**
 * Rule that requires specified directories to exist.
 */
object DirectoryPresenceRule : HarnessCheckRule() {
    /**
     * Category key.
     */
    override val category: String = "directoryPresence"

    override fun validate(ctx: RuleContext): Collection<Finding> =
        buildList {
            val catObj = ctx.manifest.categoryObject(category)
            val parametersObj = catObj?.get("parameters")?.jsonObject
            if (catObj != null && parametersObj != null) {
                val paths = ctx.manifest.stringArray(category, "paths")
                for (path in paths) {
                    val p = ctx.root / path
                    when {
                        p.isSymbolicLink() -> {
                            add(
                                Finding(
                                    Severity.ERROR,
                                    category,
                                    "symlink directory is not allowed: $path",
                                ),
                            )
                        }

                        !p.isDirectory() -> {
                            add(
                                Finding(
                                    ctx.manifest.severityOf(category),
                                    category,
                                    "missing directory: $path",
                                ),
                            )
                        }
                    }
                }
            }
        }
}
